spec.emu

```diff
@@ -30,30 +30,46 @@ contributors: "Jack Works"
 <emu-clause id="sec-algor">
     <h1>Algorithms</h1>
     <emu-clause id="sec-create-numeric-range-iterator">
-        <h1><dfn>CreateNumericRangeIterator</dfn>(_start_, _end_, _option_, _type_)</h1>
+        <h1><dfn>CreateNumericRangeIterator</dfn>(_arg0_, _arg1_, _arg2_, _type_)</h1>
         <emu-alg>
-            1. If Type(_start_) is not _type_, throw a *TypeError* exception.
             1. Assert: _type_ is *"number"* or *"bigint"*.
             1. If _type_ is *"bigint"*, let _zero_ be *0n*, else let _zero_ be *0*.
             1. If _type_ is *"bigint"*, let _one_ be *1n*, else let _one_ be *1*.
-            1. Note: <emu-note type="editor">Allowing all kinds of number (number, bigint, decimals, ...) to range from a finite number to infinity.</emu-note>
-            1. If _end_ is not *+∞* or *-∞* and if Type(_end_) is not _type_, throw a *TypeError* exception.
-            1. If _start_ is *+∞* or *-∞*, throws a *RangeError* exception.
-            1. Let _ifIncrease_ be end > start.
+            <!-- prepare -->
+            1. Let _option_ be *undefined*.
+            1. Let _start_ be *undefined*.
+            1. Let _end_ be *undefined*.
+            1. Let _step_ be *undefined*.
             1. Let _inclusiveEnd_ be *false*.
-            <!-- Normalize options -->
-            1. If _option_ is *undefined* or *null*, let _step_ be *undefined*.
-            1. Else if Type(_option_) is *object*,
-                1. Let _step_ be Get(_option_, "step").
-                1. Let _inclusiveEnd_ be ToBoolean(Get(_option_, "inclusive")).
-            1. Else if Type(_option_) is _type_, let _step_ be _option_.
+            1. If Type(_arg0_) is *"object"*,
```

Comment:
Overloading argument signatures like this is very much not a pattern the language uses, and is also one that has fallen severely out of favor in the ecosystem.

Reply:
I don't think this is a bad overload

Reply:
It’s an overload at all, which is the problem.

Reply:
so what's wrong with the overload? it can make reuse easier other wise you need to write

```js
const x = Number.range(0, 2)
consume_iterator(x)
const y = Number.range(x.start, x.end, x)
// a fresh iterator
```

Reply:
A possible alternative to achieve the same goal without overloading the `Number.range` signature
```javascript
const x = Number.range(0, 2)
consume_iterator(x)
const y = x.restart(); // bike-shed the method name
// a fresh iterator
```

Reply:
That is also a viable way to do this. I'm happy to add it if there is a suitable name.

Reply:
Name ideas
- `restart`
- `reset`
- `clone`, with a `restart` or `reset` option
```javascript
const y = x.clone( { restart: true } );
```

Reply:
I have to say all of them are strange especially it seems you'd like to preserve the current internal iterator state while cloning though I think it might be useful (like cloning a Stream)

Reply:
I have just added `reiterate` to the list. It seems, to me, that `reiterate` would successfully communicate the semantics.

Reply:
Why would you need to write that? The only form of reuse that is needed is “wrap it in a function”. Before trying to make other forms of reuse easier, it seems like a compelling case should be made that “a function” is insufficient.

Reply:
Example case 1: I'm getting the range from the outside world (code that I cannot control) so I want to get a fresh state in case the iterator has already used.

```js
export function size(range) {
    range = Number.range(range)
    return [...range].size()
}

const r = Number.range(0, 5)
r.next(); r.next();
size(r) // still get the correct answer
```

Example case 2: Side-effect / heavy computation expression when constructing

```js
const x = () => Number.range(0, prompt("end?", 10))
x(); x(); 
// bad

const y = prompt("end?", 10)
const z = () => Number.range(0, y)
z(); z(); 
// correct

const w = Number.range(0, prompt("end?", 10))
w; Number.range(w);
// better
```

Reply:
Iterators are categorically (in a general sense) not reusable, and a "fresh state" can not be obtained from them, so I don't think that's a reasonable expectation for users to have.

Certainly if the range iterator has a way to provide its "start" and "end" arguments, then the user could feed those back into `Number.range` and make a fresh iterator, but I don't think `Number.range` needs to make that easier by treating the iterators it produces as somehow special.

Reply:
The "correct" code in your Example 2 (stashing the side-effect-computed values in a variable then making a function that captures the values) seems obviously the right way to go about this; it's how you ensure that your side-effect-heavy input calculations don't get run multiple times *in every other instance of this in the language*. Like Jordan, I'm not sure why Ranges are special enough to require *extra* protection against side-effecty inputs. This is not an existing pattern in the language and we have no plans to extend it across the language, so without a pretty compelling reason why Range is special in this regard, we shouldn't introduce it.

For example 1, that is *extremely bad* code that we don't want to reinforce. The range already exposes all the information you need to calculate the remaining size (via `.start`, etc) without having to rerun and exhaust it, and if you write the function using those properties it won't crash or hang when the iterator is extremely long or infinite, either.

Reply:
I feel like we're stuck in a time loop. Arrow functions already provide reuse, you don't need to solve that in this proposal.

Reply:
More to the point, you shouldn’t try to solve it in this proposal, and you likely won’t be permitted to.

Reply:
ok, I want to point out I didn't "by treating the iterators it produces as somehow special."

I'm performing normal Get and Set on the object. I'm not reading its internal state (that is what I think about "special"). So the following code works as well.

```js
range({start: x, end: y})
```

Reply:
that's totally true, but then it still creates another kind of iterator that *is* somehow reusable, and *those* are what's confusing because it implies all iterators are reusable, but they are not.